pkg/gardenlet/controller/federatedseed/networkpolicy/controller.go

```diff
@@ -64,16 +66,18 @@ type Controller struct {
 }
 
 // NewController instantiates a new controller.
-func NewController(ctx context.Context, seedClient kubernetes.Interface, seedDefaultNamespaceKubeInformer kubeinformers.SharedInformerFactory, seedKubeInformerFactory kubeinformers.SharedInformerFactory, seedLogger *logrus.Entry, recorder record.EventRecorder, seedName string) *Controller {
+func NewController(ctx context.Context, seedClient kubernetes.Interface, seedDefaultNamespaceKubeInformer kubeinformers.SharedInformerFactory, seedLogger *logrus.Entry, recorder record.EventRecorder, seedName string) (*Controller, error) {
+	networkPoliciesInformer, err := seedClient.Cache().GetInformer(&networkingv1.NetworkPolicy{})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
What would be your preference on this: initialize/get informers in the controller constructor (`NewController` here) or doing it in when starting the controller in `Run` along with adding the event handlers?
In https://github.com/gardener/gardener/pull/2586/files#diff-8047290eb35b00e8dbcbaec8e091f641R82-R94 I chose the second option, but I am open to do it otherwise.
Do we want to agree on one of the options as a guideline?

Reply:
I don't have a preference and can adapt the PR to do it in the `Run` function. One advantage is that the error handling can be omitted in `NewController `.

Reply:
Well, that's a good point. Then let's do it in the `Run` function.
Can you then also move this part to the `Run` function?
https://github.com/gardener/gardener/blob/6e528916105623fee733d2b3c043385375a08761/pkg/gardenlet/controller/federatedseed/extensions/extensions.go#L78-L80

Reply:
While adapting these changes I realized that getting the informers during `run` may not be the best idea in general. 

Some `run` funcs (see [here](https://github.com/gardener/gardener/blob/c0c5a694ae28d90f06c080008f5464d161fb30d3/pkg/gardenlet/controller/factory.go#L174-L179)) are already called in their own Goroutine and handling such "setup" errors like getting an informer would require the implementation of an error channel. While I'm not inherently against error channels, I think it's not necessary for these cases because the error could have already been checked during the construction in a `new` func or similar.

Reply:
I see, that's why I used `panic` for error handling during controller start in #2586 
If the controllers are not able to start, because they can't get the necessary informers, we can't do anything anyways even with proper error handling. So I think `panic` is fine for that case.
For the federated Seed controllers OTOH, `panic` is of course no solution. But I think, we already had error handling for the `Run` funcs before, right?
I am also open to get informers it in the `new` funcs, although I think that handling the errors there, will basically lead to the same result for the "normal" controllers (-> let gardenlet exit with error code).
WDYT?

Reply:
> So I think panic is fine for that case.

We should only use a panic (if possible) when we are in a general intialization, i.e. before controllers or futher long running Goroutines are started. The panic can happen when other controllers already run and thus they are terminated ungracefully. Of course, (most) actions are idempotent but still, not very nice.

> But I think, we already had error handling for the Run funcs before, right?

Not really, because it was not necessary, except when we waited for caches to sync ([example](https://github.com/gardener/gardener/blob/f120f2c74a503ab4a388aa055cd08d4081eaf5ed/pkg/gardenlet/controller/backupbucket/backup_bucket.go#L94-L97)) and this case was more or less disregarded because it was quite unlikely to happen. I just saw that we have this panic usage [here](https://github.com/gardener/gardener/blob/c0c5a694ae28d90f06c080008f5464d161fb30d3/pkg/gardenlet/controller/factory.go#L93) as well. But for the reason I explained above, I'm not really convinced yet that this is the way to go.

If we already have the chance to detect runtime errors early during initialization why not use this advantage and exit  before the controllers might go ahead and alter the system?

Reply:
> The panic can happen when other controllers already run and thus they are terminated ungracefully.

Yes, you're right. Just thought about it again and it may indeed take some time for the calls to `GetInformer` to return an error (waiting for the cache might timeout or whatever) and then other controllers might already run. Then it would be better of course to have proper error handling. I just thought, that the errors might arise directly, but that's not the case.
So agreed :)

Let's move the informer creation/retrieval to the `new` funcs then.
And how would you like to handle waiting for the caches then? This should not happen in the `new` func, I think, because then the different controllers can't wait for the respective caches to sync in parallel and it is rather a serial and blocking initialization.
So if we want to handle that gracefully, we would have to implement proper error handling for the `run` funcs also, right?

Reply:
> And how would you like to handle waiting for the caches then?

I wouldn't touch that now and leave it the way it is done in the `run` functions today. I've never seen this causing us any trouble.

Reply:
Ok, agreed. 👍